tests/e2e/e2e_test.go

```diff
@@ -297,10 +310,90 @@ func (ctx *testCtx) pingFromVPPClb(destAddress string) func() error {
 	}
 }
 
-/*func (ctx *testCtx) testConnection(fromMs, toMs, dstAddr, listenAddr string, port uint16, udp bool) error {
-	// TODO (run nc client and server)
-	return nil
-}*/
+func (ctx *testCtx) testConnection(fromMs, toMs, toAddr, listenAddr string,
+	toPort, listenPort uint16, udp bool, traceVPPNodes ...string) error {
+
+	const (
+		connTimeout    = 3 * time.Second
+		srvExitTimeout = 500 * time.Millisecond
+		reqData        = "Hi server!"
+		respData       = "Hi client!"
+	)
+
+	clientMs, found := ctx.microservices[fromMs]
+	if !found {
+		// bug inside a test
+		ctx.t.Fatalf("cannot run TCP/UDP client from unknown microservice '%s'", fromMs)
```

Comment:
The function `(*testing.T).Helper` might help when this fails. It will show line number for the caller instead of this line.